File: packages/language-adapters/src/languageDetector.ts
// Automatic Language Detector for CodeFlow Platform

export type SupportedLanguage = 'python' | 'javascript' | 'typescript' | 'cpp' | 'java';

export interface ILanguageDetectionResult {
  language: SupportedLanguage;
  confidence: number;
  evidence: string[];
}

export class LanguageDetector {
  /**
   * Automatically detect programming language from code syntax patterns
   */
  static detectLanguage(code: string): ILanguageDetectionResult {
    if (!code || !code.trim()) {
      return { language: 'python', confidence: 1.0, evidence: ['Default fallback'] };
    }

    const trimmed = code.trim();
    const scores: Record<SupportedLanguage, number> = {
      python: 0,
      javascript: 0,
      typescript: 0,
      cpp: 0,
      java: 0,
    };
    const evidenceMap: Record<SupportedLanguage, string[]> = {
      python: [],
      javascript: [],
      typescript: [],
      cpp: [],
      java: [],
    };

    // Python signals
    if (/\bdef\s+\w+\s*\(/.test(trimmed)) { scores.python += 30; evidenceMap.python.push('def keyword'); }
    if (/\bimport\s+\w+|\bfrom\s+\w+\s+import/.test(trimmed)) { scores.python += 20; evidenceMap.python.push('import/from statement'); }
    if (/\belif\b/.test(trimmed)) { scores.python += 25; evidenceMap.python.push('elif keyword'); }
    if (/\bNone\b|\bTrue\b|\bFalse\b/.test(trimmed)) { scores.python += 20; evidenceMap.python.push('Python booleans/None'); }
    if (/\bself\.\w+/.test(trimmed)) { scores.python += 25; evidenceMap.python.push('self attribute access'); }
    if (/\bprint\s*\(/.test(trimmed) && !/System\.out/.test(trimmed)) { scores.python += 15; evidenceMap.python.push('print() function'); }
    if (/\bset\(\)|\bdeque\(|\bheapq\./.test(trimmed)) { scores.python += 25; evidenceMap.python.push('Python stdlib data structures'); }

    // C++ signals
    if (/#include\s*<.+>/.test(trimmed)) { scores.cpp += 40; evidenceMap.cpp.push('#include directive'); }
    if (/\bstd::/.test(trimmed)) { scores.cpp += 35; evidenceMap.cpp.push('std:: namespace'); }
    if (/\bcout\s*<</.test(trimmed)) { scores.cpp += 35; evidenceMap.cpp.push('cout stream'); }
    if (/\bvector\s*</.test(trimmed)) { scores.cpp += 30; evidenceMap.cpp.push('std::vector template'); }
    if (/\bint\s+main\s*\(/.test(trimmed)) { scores.cpp += 30; evidenceMap.cpp.push('main function'); }

    // Java signals
    if (/\bpublic\s+class\s+\w+/.test(trimmed)) { scores.java += 40; evidenceMap.java.push('public class definition'); }
    if (/\bpublic\s+static\s+void\s+main/.test(trimmed)) { scores.java += 40; evidenceMap.java.push('main method'); }
    if (/System\.out\.println/.test(trimmed)) { scores.java += 35; evidenceMap.java.push('System.out.println'); }
    if (/\bimport\s+java\./.test(trimmed)) { scores.java += 30; evidenceMap.java.push('import java.*'); }

    // JS/TS signals
    if (/\bconst\s+|\blet\s+|\bvar\s+/.test(trimmed)) {
      scores.javascript += 20; evidenceMap.javascript.push('const/let/var decl');
      scores.typescript += 20; evidenceMap.typescript.push('const/let/var decl');
    }
    if (/\bfunction\s+\w+\s*\(/.test(trimmed)) {
      scores.javascript += 20; evidenceMap.javascript.push('function keyword');
      scores.typescript += 20; evidenceMap.typescript.push('function keyword');
    }
    if (/=>/.test(trimmed)) {
      scores.javascript += 15; evidenceMap.javascript.push('arrow function');
      scores.typescript += 15; evidenceMap.typescript.push('arrow function');
    }

    // TypeScript specific signals
    if (/\binterface\s+\w+|\btype\s+\w+\s*=/.test(trimmed)) { scores.typescript += 35; evidenceMap.typescript.push('interface/type definition'); }
    if (/:\s*(number|string|boolean|any|void)\b/.test(trimmed)) { scores.typescript += 30; evidenceMap.typescript.push('type annotations'); }

    // Find highest score
    let bestLang: SupportedLanguage = 'python';
    let maxScore = scores.python;

    (Object.keys(scores) as SupportedLanguage[]).forEach(lang => {
      if (scores[lang] > maxScore) {
        maxScore = scores[lang];
        bestLang = lang;
      }
    });

    const confidence = Math.min(0.98, Math.max(0.60, maxScore / 50));

    return {
      language: bestLang,
      confidence,
      evidence: evidenceMap[bestLang],
    };
  }
}

File: services/backend/app/utils/languageDetector.py
import re
from typing import Dict, Any

class LanguageDetector:
    """
    Backend Language Auto-Detector using syntax heuristics, keyword patterns, and idiom scoring.
    Returns { "language": str, "confidence": float, "dialect": str }
    """

    PATTERNS = {
        "python": [
            (r'\bdef\s+\w+\s*\(', 30),
            (r'\bimport\s+\w+|\bfrom\s+\w+\s+import', 25),
            (r'\bif\s+__name__\s*==\s*[\'"]__main__[\'"]:', 40),
            (r'\bprint\s*\(', 15),
            (r'\bself\b', 20),
            (r':\s*$', 15),
        ],
        "javascript": [
            (r'\bconst\s+|\blet\s+|\bvar\s+', 25),
            (r'\bfunction\s+\w+\s*\(', 25),
            (r'\bconsole\.log\s*\(', 30),
            (r'=>', 20),
            (r'\bmodule\.exports\b|\bexport\s+default\b', 30),
        ],
        "typescript": [
            (r':\s*(number|string|boolean|any|void|object)\b', 35),
            (r'\binterface\s+\w+', 35),
            (r'\btype\s+\w+\s*=', 30),
        ],
        "cpp": [
            (r'#include\s*<\w+>', 40),
            (r'\bstd::\w+', 35),
            (r'\bcout\s*<<|\bcin\s*>>', 40),
            (r'\bint\s+main\s*\(\s*\)', 35),
            (r'struct\s+\w+\s*\{', 20),
        ],
        "java": [
            (r'\bpublic\s+class\s+\w+', 40),
            (r'\bpublic\s+static\s+void\s+main\s*\(', 45),
            (r'\bSystem\.out\.println\s*\(', 40),
            (r'\bimport\s+java\.', 35),
        ],
        "go": [
            (r'\bpackage\s+main\b', 45),
            (r'\bfunc\s+main\s*\(\s*\)', 45),
            (r'\bfmt\.Println\s*\(', 40),
            (r'\bimport\s*\(\s*"fmt"', 40),
        ],
        "rust": [
            (r'\bfn\s+main\s*\(\s*\)', 45),
            (r'\blet\s+mut\b', 40),
            (r'\bprintln!\s*\(', 40),
            (r'\buse\s+std::', 35),
        ]
    }

    @classmethod
    def detect(cls, code: str) -> Dict[str, Any]:
        if not code or not code.strip():
            return {"language": "python", "confidence": 0.5, "dialect": "standard"}

        scores = {lang: 0 for lang in cls.PATTERNS.keys()}

        for lang, pattern_list in cls.PATTERNS.items():
            for pattern, weight in pattern_list:
                if re.search(pattern, code, re.MULTILINE):
                    scores[lang] += weight

        # TypeScript inherits JS score
        scores["typescript"] += scores["javascript"] // 2

        best_lang = max(scores, key=scores.get)
        best_score = scores[best_lang]

        if best_score == 0:
            return {"language": "python", "confidence": 0.5, "dialect": "standard"}

        confidence = min(best_score / 100.0, 0.99)
        return {
            "language": best_lang,
            "confidence": round(confidence, 2),
            "dialect": "standard"
        }
